6.0 Update - Web API Migration (#24601)
dotnet/AspNetCore.Docs: aspnetcore/migration/webapi.md

description: Learn how to migrate a web API implementation from ASP.NET 4.x Web API to ASP.NET Core MVC.
ms.author: scaddie
ms.custom: mvc
ms.date: 01/31/2022
no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: migration/webapi
---
# Migrate from ASP.NET Web API to ASP.NET Core

> An ASP.NET 4.x Web API is an HTTP service that reaches a broad range of clients, including browsers and mobile devices. ASP.NET Core combines ASP.NET 4.x's MVC and Web API app models into a single programming model known as ASP.NET Core MVC.

Author: I will delete the blurb on 4.x.  I was trying to show the relationship, not promote it of course, but I understand your point.

:::moniker range=">= aspnetcore-6.0"
This article migrates a basic controller from ASP.NET 4.x to to ASP.NET Core 6.x.

> [View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/migration/webapi/sample/) ([how to download](xref:index#how-to-download-a-sample))

Author: I kept the link to the sample project at the top for this one since that is the sample we are referring to that we are migrating for this doc.  It seemed worth providing a link at the top to save confusion.  I will move it to the bottom however.

1. Right-click the **Models** folder. Select **Add** > **Class**. Name the class *Product* and select **Add**.
1. Replace the template model code with the following:

   [!code-csharp[](webapi/sample/3.x/ProductsApp/Models/Product.cs)]

Add the example controller:

1. Right-click the **Controllers** folder.
1. Select **Add > Controller...**.
1. In **Add New Scaffolded Item** dialog, select **Mvc Controller - Empty** then select **Add**.
1. Name the controller *ProductsController* and select **Add**.
1. Replace the template controller code with the following:

   [!code-csharp[](webapi/sample/3.x/ProductsApp/Controllers/ProductsController.cs)]

## The configuration for ASP.NET Core Web API

ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file. The *web.config* file is added at publish time.

The *Program.cs* file:

* Replaces *Global.asax*.
* Handles all app startup tasks.

For more information, see <xref:fundamentals/startup>.

The following shows the application startup code in the ASP.NET Core *Program.cs* file:

[!code-csharp[](webapi/sample/6.x/ProductsCore/Program.cs)]

> ## Update the model

Author: Thanks, I misunderstood what you were asking for originally and removed the explanation and linked out.

    Attribute routing supports tokens, such as `[controller]` and `[action]`. At runtime, each token is replaced with the name of the controller or action, respectively, to which the attribute has been applied. The tokens:
    * Reduce the number of magic strings in the project.
    * Ensure routes remain synchronized with the corresponding controllers and actions when automatic rename refactorings are applied.
1. Enable HTTP Get requests to the `ProductController` actions:
    * Apply the [`[HttpGet]`](xref:Microsoft.AspNetCore.Mvc.HttpGetAttribute) attribute to the `GetAllProducts` action.
    * Apply the `[HttpGet("{id}")]` attribute to the `GetProduct` action.

Run the migrated project, and browse to `/api/products`.  For example: https://localhost:`<port>`/api/products. A full list of three products appears. Browse to `/api/products/1`. The first product appears.

## Additional resources

* <xref:web-api/index>
* <xref:web-api/action-return-types>
* <xref:mvc/compatibility-version>

:::moniker-end

> :::moniker range="< aspnetcore-5.0 >= aspnetcore-3.0"

Author: The original covered:
:::moniker range=">= aspnetcore-3.0"
:::moniker range="<= aspnetcore-2.2"

We would keep those versions 3.x and 2.x   wouldn't we?

> :::moniker range="< aspnetcore-5.0 >= aspnetcore-3.0"

Reviewer: Keep 3.1 and higher.

> :::moniker range="< aspnetcore-5.0 >= aspnetcore-3.0"

Author: We are actively tossing 2.x?  I feel uncomfortable deleting those with no archiving system in place.  How are we keeping 2.x?

> :::moniker range="< aspnetcore-5.0 >= aspnetcore-3.0"

Reviewer: We're not keeping out of support versions. It doesn't make any sense to keep out of support versions for a migration guide. Why would anyone decide to migrate from ASP.NET 4.x to ASP.NET Core 2.x?
See https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

description: Learn how to migrate a web API implementation from ASP.NET 4.x Web API to ASP.NET Core MVC.
ms.author: scaddie
ms.custom: mvc
ms.date: 01/31/2022
no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: migration/webapi
---
# Migrate from ASP.NET Web API to ASP.NET Core

An ASP.NET 4.x Web API is an HTTP service that reaches a broad range of clients, including browsers and mobile devices. ASP.NET Core combines ASP.NET 4.x's MVC and Web API app models into a single programming model known as ASP.NET Core MVC. This article demonstrates the steps required to migrate from ASP.NET 4.x Web API to ASP.NET Core MVC.

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/migration/webapi/sample) ([how to download](xref:index#how-to-download-a-sample))

::: moniker range=">= aspnetcore-6.0"

## Prerequisites

[!INCLUDE [prerequisites](../includes/net-core-prereqs-vs-3.1.md)]

## Review ASP.NET 4.x Web API project

> This article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Author: "No need to have them bring that".    What do you mean by "that".  No need to bring in the previous project?  Sorry could you be more explicit?

> This article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Reviewer: Don't bring in the old project. They just need the model and controller.

> This article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Author: Sorry, one more question to make sure I am on the same page.  So just keep to migrating the controller, the model it uses and program.cs to show where routing is enabled.  I also pull out the section: _Review ASP.NET 4.x Web API project_ which includes mention of the original WebAPIConfig.cs attribute routing.  Do you want to compare old routing to new routing or just stick with how the new routing is set up?  Thanks for all the help.

> This article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Reviewer: Correct. No need to review the project, those folks are experts in 4.x web api.
No need to compare old routing to new routing, provide links to routing and creating API content.

> This article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Author: OK, it sounds like I will just instruct to create a new ASP.NET Core Web API project rather than the blank solution first.  I am not to show the old controller or model (nor the whole previous project), but rather just show the aspects of what a new controller and model looks like and what is happening with that.    That feels really narrow compared to the title.  Should I change the title to be specific to migrating a 4.x controller rather than a 4.x Web API project?   Sorry for all the questions.

* Configures the routing table.
The sample code expects URLs to match the format `/api/{controller}/{id}`, with `{id}` being optional.

The following sections demonstrate migration of the Web API project to ASP.NET Core MVC.

## Create the destination project

### Prerequisites

[!INCLUDE[](~/includes/net-prereqs-vs-6.0.md)]

Create a new blank solution in Visual Studio and add the ASP.NET 4.x Web API project to migrate:

1. From the **File** menu, select **New** > **Project**.
1. Select the **Blank Solution** template and select **Next**.
1. Name the solution *WebAPIMigration*. Select **Create**.
1. Add the existing *ProductsApp* project to the solution.

Add a new API project to migrate to:

1. Add a new **ASP.NET Core Web API** project to the solution, and select **Next**.
1. In the **Configure your new project** dialog, name the project *ProductsCore*, and select **Next**.
1. In the **Additional information** dialog:
   1. Confirm the **Framework** is **.NET 6.0 (Long-term support)**.
   1. Confirm the checkbox for **Use controllers(uncheck to use minimal APIs)** is checked.
   1. Uncheck **Enable OpenAPI support**.
   1. Select **Create**.

> 1. Remove the *WeatherForecast.cs* and *Controllers/WeatherForecastController.cs* example files from the new *ProductsCore* project.

Reviewer: ```suggestion
1. Remove the *WeatherForecast.cs* and *Controllers/WeatherForecastController.cs* example files from the new *ProductsCore* project.
something like
1. Copy the ASP.NET 4.x Web API *ProductsController * and model to the new project.
```

1. Open *Properties\launchSettings.json*.
1. Change `launchUrl` properties from `weatherforcast` to `productscore`.

The solution now contains two projects. The following sections explain migrating the *ProductsApp* project's contents to the *ProductsCore* project.

## Migrate configuration

ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file. The *web.config* file is added at publish time.

The *Program.cs* file:

* Replaces *Global.asax*.
* Handles all app startup tasks.

For more information, see <xref:fundamentals/startup>.

The following shows the application startup code in the ASP.NET Core *Program.cs* file:

[!code-csharp[](webapi/sample/6.x/ProductsCore/Program.cs)]

## Migrate the model

> 

Reviewer 2: Yes, I agree.

ASP.NET Core 6.0 projects enable nullable reference types by default. Nullable reference types are enabled or disabled for the entire project with the following project files property:

```xml
<PropertyGroup>
    <Nullable>enable</Nullable>
</PropertyGroup>
```

For more information, see [Nullable reference types](/dotnet/csharp/nullable-references).

Examine *Models/Product.cs*. With nullable reference types enabled for the ProductCore project, helpful `non-nullable property` warnings are visible for the `Name` and `Category` properties.

![Nullable reference type warning](webapi/_static/non-nullable-warning.png)

In this case, the intent is that the `Name` and `Category` can be nullable types.

1. Add the `?` annotation to declare the `Name` and `Category` properties as nullable types:

```diff
public int Id { get; set; }
- public string Name { get; set; }
+ public string? Name { get; set; }
- public string Category { get; set; }
+ public string? Category { get; set; }
public decimal Price { get; set; }
```

## Migrate the controller

Update the `ProductsController` for ASP.NET Core:

> The following code shows the updated ASP.NET Core `ProductsController`:

Reviewer: ```suggestion
```

> [!code-csharp[](webapi/sample/6.x/ProductsCore/Controllers/ProductsController.cs?highlight=1,2,4,8,33,50,57)]

Reviewer: Highlight 
```
[Route("api/[controller]")]
[ApiController]
```

1. Copy *Controllers/ProductsController.cs* and the *Models* folder from the original project to the new one.
1. Change the copied files' root namespace to `ProductsCore`.
1. Update the `using ProductsApp.Models;` statement to `using ProductsCore.Models;`.

The following components don't exist in ASP.NET Core:

* `ApiController` class
* `System.Web.Http` namespace
* `IHttpActionResult` interface

> Make the following changes:

Reviewer: ```suggestion
```

1. Change `ApiController` to <xref:Microsoft.AspNetCore.Mvc.ControllerBase>. Add `using Microsoft.AspNetCore.Mvc;` to resolve the `ControllerBase` reference.
1. Delete `using System.Web.Http;`.

> 1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`.

Reviewer: ```suggestion
1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`. For more info, see https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-6.0
```

1. Simplify the `GetProduct` action's `return` statement to the following statement:

> 

Reviewer: ```
[Route("api/[controller]")]
[ApiController]
```
is explained in the next section.

    ```csharp
    return product;
    ```

> ## Configure routing

Reviewer: ```suggestion
## Routing
```

> The following code code in the App_Start/WebAPIConfig.cs added an endpoint for the ASP.NET Core 3.x ProductsApp app:

Reviewer: ```suggestion
The ASP.NET Web API 2 *App_Start/WebAPIConfig.cs* file contains an endpoint for the ProductsApp app:
```

[!code-csharp[](webapi/sample/3.x/ProductsApp/App_Start/WebAPIConfig.cs?highlight=15, 17-21)]

> ASP.NET Core 6.0 provides a minimal hosting model in which the endpoint routing middleware wraps the entire middleware pipeline, therefore routes can be added directly to the <xref:Microsoft.AspNetCore.Builder.WebApplication> without an explicit call to <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A> or <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> to register routes.

Reviewer: We almost never put version numbers in the doc. This is the V6 version and won't be updated when V7 is released.

Author: @Rick-Anderson .   It is helpful I think to know when the switch was made.  Perhaps something like "ASP.NET Core 6.0 introduced a minimal hosting model..." ?

Reviewer: No, doesn't meet the bar. If they're on V6, we don't care about history.

Author: I disagree.  This is a migration doc, not just any doc.  I will make the change anyway, but I think it needed to be explicit.

> ASP.NET Core 6.0 provides a minimal hosting model in which the endpoint routing middleware wraps the entire middleware pipeline, therefore routes can be added directly to the <xref:Microsoft.AspNetCore.Builder.WebApplication> without an explicit call to <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A> or <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> to register routes.

Reviewer: ```suggestion
ASP.NET Core provides a minimal hosting model in which the endpoint routing middleware wraps the entire middleware pipeline, therefore routes can be added directly to the <xref:Microsoft.AspNetCore.Builder.WebApplication> without an explicit call to <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A> or <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> to register routes.
```

`UseRouting` can still be used to specify where route matching happens, but `UseRouting` doesn't need to be explicitly called if routes should be matched at the beginning of the middleware pipeline.

> In *Program.cs* in the ASP.NET Core 6.0 ProductsCore app, <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllers%2A> is called inside `UseEndpoints` to map attribute routed controllers.

Reviewer: Fix

[!code-csharp[](webapi/sample/6.x/ProductsCore/Program.cs?highlight=15)]

**Note:** Routes added directly to the <xref:Microsoft.AspNetCore.Builder.WebApplication> execute at the ***end*** of the pipeline.

Configure routing as follows:

> 1. Mark the `ProductsController` class with the following highlighted attributes:

Reviewer: ```suggestion
1. The migrated `ProductsController` contains the following highlighted attributes:
```

`UseRouting` can still be used to specify where route matching happens, but `UseRouting` doesn't need to be explicitly called if routes should be matched at the beginning of the middleware pipeline.

In *Program.cs* in the ASP.NET Core ProductsCore app, <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllers%2A> is called inside `UseEndpoints` to map attribute routed controllers.

[!code-csharp[](webapi/sample/6.x/ProductsCore/Program.cs?highlight=15)]

**Note:** Routes added directly to the <xref:Microsoft.AspNetCore.Builder.WebApplication> execute at the ***end*** of the pipeline.

Configure routing as follows:

1. The migrated `ProductsController` contains the following highlighted attributes:

    [!code-csharp[](webapi/sample/6.x/ProductsCore/Controllers/ProductsController.cs?highlight=6,7)]

    The preceding [`[Route]`](xref:Microsoft.AspNetCore.Mvc.RouteAttribute) attribute configures the controller's attribute routing pattern. The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute makes attribute routing a requirement for all actions in this controller.

>     Attribute routing supports tokens, such as `[controller]` and `[action]`. At runtime, each token is replaced with the name of the controller or action, respectively, to which the attribute has been applied. The tokens:

Author: Do you mean link them both to [Token replacement in route templates [controller], [action], [area]](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?#token-replacement-in-route-templates-controller-action-area).   That is the only place I know of that we define them well.  If you mean they should be linked to API reference, I honestly don't know where that would be in the reference or if there is one exposed that matches.  Sorry to ask.

>     Attribute routing supports tokens, such as `[controller]` and `[action]`. At runtime, each token is replaced with the name of the controller or action, respectively, to which the attribute has been applied. The tokens:

Reviewer: make `Attribute routing supports tokens` a link to [Token replacement in route templates [controller], [action], [area]](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?#token-replacement-in-route-templates-controller-action-area).

>     * Reduce the number of magic strings in the project.

Reviewer: ```suggestion
    * Reduces or eliminates the need to use hard coded strings for the route.
```

Add a new API project to migrate to:

1. Add a new **ASP.NET Core Web API** project to the solution, and select **Next**.
1. In the **Configure your new project** dialog, name the project *ProductsCore*, and select **Next**.
1. In the **Additional information** dialog:
  * Confirm the **Framework** is **.NET 6.0 (Long-term support)**.
  * Confirm the checkbox for **Use controllers(uncheck to use minimal APIs)** is checked.

>   * Select **Create**.

Reviewer: Uncheck Open API

ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file. Additionally, the *web.config* file is added at publish time.

> The `Program` class:

Reviewer: It's not a class.

1. Change `ApiController` to <xref:Microsoft.AspNetCore.Mvc.ControllerBase>. Add `using Microsoft.AspNetCore.Mvc;` to resolve the `ControllerBase` reference.
1. Delete `using System.Web.Http;`.
1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`.
1. Simplify the `GetProduct` action's `return` statement to the following statement:

    ```csharp
    return product;
    ```

The following code shows the updated ASP.NET Core `ProductsController`:

[!code-csharp[](webapi/sample/6.x/ProductsCore/Controllers/ProductsController.cs?highlight=1,2,8,12,37,44)]

> ## Configure routing

Reviewer: Everything in routing is wrong.
